lib/hlclock.ex

```diff
@@ -1,50 +1,104 @@
+  clock at rest will eventually be unable to generate timestamps due to assumed
+  bounds on the logical clock relative to physical time.
+
+  In order to account for physical time drift within the system, timestamps
+  should regularly be exchanged between nodes. Generate a timestamp at one node
+  via HLClock.send_timestamp/1; at the other node, call HLClock.recv_timestamp/2
+  with the received timestamp from the first node.
+
+  Inspired by https://www.cse.buffalo.edu/tech-reports/2014-04.pdf
   """
 
+  alias __MODULE__, as: Clock
   alias HLClock.Timestamp
 
-  def new(pt \\ default_clock()) do
-    Timestamp.new(pt, 0)
+  defstruct [:clock_fn, :curr_ts]
+
+  @doc """
+  clock constructor requires the node_id, a millisecond clock fn and a
+  maximum drift parameter in milliseconds
+  """
+  def new(node_id \\ 0,
+    clock_fn \\ &Clock.default_time/0) do
+    with {:ok, t0} <- Timestamp.new(clock_fn.(), 0, node_id) do
+      %Clock{clock_fn: clock_fn,
+             curr_ts: t0}
+    end
   end
 
-  def now(clock, pt \\ default_clock()) do
-    if Timestamp.ahead?(clock, pt) do
-      Timestamp.increment(clock)
-    else
-      Timestamp.new(pt, 0)
+  @doc "generate a new timestamp"
+  def send_timestamp(%Clock{curr_ts: old, clock_fn: cfn} = clock) do
+    pt = cfn.()
+    nt = max(old.time, pt)
+    nc = advance(old, nt)
+
+    cond do
+      drift?(clock, nt, pt) ->
+        {:error, :clock_drift_violation}
+      true ->
+        # counter overflow is handled by Timestamp.new
+        with {:ok, new_ts} <- Timestamp.new(nt, nc, old.node_id) do
+          {:ok, %Clock{clock | curr_ts: new_ts}}
+        end
     end
   end
 
-  def less?(clock1, clock2) do
-    Timestamp.compare(clock1, clock2) == :lt
+  defp advance(old, new_time) do
+    cond do
+      old.time == new_time ->
+        old.counter + 1
+      true ->
+        0
+    end
   end
 
-  def update(local, foreign, pt \\ default_clock()) do
-    max_pt = max(pt, max(local.time, foreign.time))
+  @doc "receive a remote timestamp and merge with local"
+  def recv_timestamp(%Clock{clock_fn: cfn, curr_ts: local} = clock, msg) do
+    pt = cfn.()
+    max_pt = max(pt, max(msg.time, local.time))
     cond do
-      max_pt == local.time && max_pt == foreign.time ->
-        max_logical = max(local.logical, foreign.logical) + 1
-        Timestamp.new(max_pt, max_logical)
-      foreign.time > local.time && remote_drifted?(foreign, pt) ->
-        local
-      max_pt == local.time ->
-        Timestamp.new(max_pt, local.logical+1)
-      max_pt == foreign.time ->
-        Timestamp.new(max_pt, foreign.logical+1)
+      local.node_id == msg.node_id ->
+        {:error, :duplicate_node_id}
+      drift?(msg.time, pt) ->
+        {:error, :remote_drift_violation}
+      drift?(max_pt, pt) ->
+        {:error, :clock_drift_violation}
       true ->
-        Timestamp.new(max_pt, 0)
+        new_log = merge_logical(max_pt, local, msg)
```

Comment:
Yeah `with` is my favorite thing they've added to the language. If your functions return meaningful `{:error, error.t}` tuples then you can just let the errors fall through. If you want to wrap those errors into something more meaningful (for instance if you're dealing with a dependency and don't wanna return their error) you can do something like this:

```elixir
with {:ok, clock} <- handle_drift(clock),
        {:ok, result} <- do_some_thing(clock) do
else
  {:error, :some_error} ->
    {:error, :failed_to_build_clock}
end
```